lib/src/Pos/Util/UserPublic.hs

```diff
@@ -0,0 +1,284 @@
+#endif
+
+-- | Create user public file at the given path, but only when one doesn't
+-- already exist.
+#ifdef POSIX
+initializeUserPublic :: (MonadIO m, WithLogger m) => FilePath -> m ()
+#else
+initializeUserPublic :: (MonadIO m) => FilePath -> m ()
+#endif
+initializeUserPublic publicPath = do
+    exists <- liftIO $ doesFileExist publicPath
+#ifdef POSIX
+    if exists
+    then ensureModeIs600 publicPath
+    else do
+        createEmptyFile publicPath
+        setMode600 publicPath
+#else
+    unless exists $ createEmptyFile publicPath
+#endif
+  where
+    createEmptyFile :: (MonadIO m) => FilePath -> m ()
+    createEmptyFile = liftIO . flip writeFile mempty
+
+-- | Reads user public from file, assuming that file exists,
+-- and has mode 600, throws exception in other case
+#ifdef POSIX
+readUserPublic :: (MonadIO m, WithLogger m) => FilePath -> m UserPublic
+#else
+readUserPublic :: (MonadIO m) => FilePath -> m UserPublic
+#endif
+readUserPublic path = do
+#ifdef POSIX
+    ensureModeIs600 path
+#endif
+    takeReadLock path $ do
+        content <- either (throwM . UserPublicDecodingError . toText) pure .
+                   decodeFull' =<< BS.readFile path
+        pure $ content & upPath .~ path
+
+-- | Reads user public from the given file.
+-- If the file does not exist/is empty, returns empty user public
+#ifdef POSIX
+peekUserPublic :: (MonadIO m, WithLogger m) => FilePath -> m UserPublic
+#else
+peekUserPublic :: (MonadIO m) => FilePath -> m UserPublic
+#endif
+peekUserPublic path = do
+    initializeUserPublic path
+    takeReadLock path $ do
+        content <- decodeFull' <$> BS.readFile path
+        pure $ either (const def) identity content & upPath .~ path
+
+-- | Read user public putting an exclusive lock on it. To unlock, use
+-- 'writeUserPublicRelease'.
+#ifdef POSIX
+takeUserPublic :: (MonadIO m, WithLogger m) => FilePath -> m UserPublic
+#else
+takeUserPublic :: (MonadIO m) => FilePath -> m UserPublic
+#endif
+takeUserPublic path = do
+    initializeUserPublic path
+    liftIO $ do
+        lock <- lockFile (lockFilePath path) Exclusive
+        content <- decodeFull' <$> BS.readFile path
+        pure $ either (const def) identity content
+            & upPath .~ path
+            & upLock .~ Just lock
+
+-- | Writes user public.
+writeUserPublic :: (MonadIO m) => UserPublic -> m ()
+writeUserPublic up
+    | canWrite up = liftIO $ throwM UserPublicAlreadyLocked
+    | otherwise   = liftIO $ withFileLock (lockFilePath $ up ^. upPath) Exclusive $ const $ writeRaw up
+
+-- | Writes user public and releases the lock. UserPublic can't be
+-- used after this function call anymore.
+writeUserPublicRelease :: (MonadIO m, MonadThrow m) => UserPublic -> m ()
+writeUserPublicRelease up
+    | not (canWrite up) = throwM UserPublicNotWritable
+    | otherwise = liftIO $ do
+        writeRaw up
+        case (up ^. upLock) of
+            Nothing   -> throwM UserPublicIncorrectLock
+            Just lock -> unlockFile lock
+
```

Comment:
Ok, we have a lock with separate take and release operations that are exported via the public API. I am wary of this pattern because its relatively easy to take the lock and forget to release if or to remember to release it but the release is accidentally dropped from the code when other work is done on it. We also have exceptions flying around here. Is it possible for the lock to be taken and then have an exception cause the release to not happen?

It would be good to investigate the possibilty of having a `withUserPublicLock` function that acts like `bracket` from `Control.Exception` and then hiding the explicit take/release mechanism.